packages/@o3r/rules-engine/src/engine/ruleset-executor.ts

```diff
@@ -195,38 +195,62 @@ export class RulesetExecutor {
     throw new Error(`Unknown condition block met : ${JSON.stringify(nestedCondition)}`);
   }
 
+  // TODO: make public and move to top of the class in a separate PR to make the diff more meaningful in this one
   /**
    * Plug ruleset to fact streams and trigger a first evaluation
-   *
-   * @param ruleset
    */
   public prepareRuleset() {
-    const rulesWithContext: Rule[] = [];
-    const rulesWithoutContext: Rule[] = [];
+    const inputFactsForRule: Record<string, string[]> = {};
+    this.ruleset.rules.forEach((rule) => {
+      const findRuleInputFacts = (obj: AllBlock): string[] => {
+        const ruleInputFacts: string[] = [];
+        // Recursive function to explore each object
+        const exploreObject = (currentObj: any) => {
```

Comment:
I'd be interested in having a check with a a big enough rulesets file to see how long it takes to be executed (you can just duplicate the IT tests json content).

Reply:
time with 1000 rules: 3.7748750001192093

total test time: 4.24 s

time with 10000: 38.763249998912215

total test time: 24.12s